extensions/positron-python/src/client/llm-tools/llm-tools.ts

```diff
@@ -0,0 +1,119 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (C) 2025 Posit Software, PBC. All rights reserved.
+ *  Licensed under the Elastic License 2.0. See LICENSE.txt for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as vscode from 'vscode';
+import { IServiceContainer } from '../../client/ioc/types';
+import { IInterpreterService } from '../../client/interpreter/contracts';
+import { IPythonExecutionFactory } from '../../client/common/process/types';
+
+export function registerPythonLanguageModelTools(
+    context: vscode.ExtensionContext,
+    serviceContainer: IServiceContainer,
+): void {
+    const pythonLoadedPackagesTool = vscode.lm.registerTool<{}>('getAttachedPythonPackages', {
+        invoke: async (_options, _token) => {
+            const interpreterService = serviceContainer.get<IInterpreterService>(IInterpreterService);
+            const activeInterpreter = await interpreterService.getActiveInterpreter();
+            if (!activeInterpreter) {
+                return new vscode.LanguageModelToolResult([
+                    new vscode.LanguageModelTextPart('No active Python interpreter'),
+                ]);
+            }
+
+            const pythonExecutionFactory = serviceContainer.get<IPythonExecutionFactory>(IPythonExecutionFactory);
+            const pythonExecution = await pythonExecutionFactory.createActivatedEnvironment({
+                interpreter: activeInterpreter,
+            });
+
+            const script = `
+import json
+import importlib.metadata
+installed_packages = [f"{dist.metadata['Name']} ({dist.version})" for dist in importlib.metadata.distributions()]
+print(json.dumps(installed_packages))
```

Comment:
This is in the loaded/attached tool, but is querying for all installed packages. In my testing it made Assistant think that every installed package was also loaded.

Reply:
Thank you! I rewrote the tool to use callMethod and added a new rpc to get the loaded modules.